### PR TITLE
Dramatically improve pod validation

### PR DESCRIPTION
**Pull request by jasongilanfarr**
Summary: - pods validation was missing a ton of stuff

Test Plan: sbt test

Reviewers: jdef, aquamatthias, unterstein

Subscribers: marathon-team

Differential Revision: https://phabricator.mesosphere.com/D29

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala [hunk 1]
```diff
@@ -64,9 +65,84 @@ trait PodsValidation {
...
+      case ShellCommand(shell) =>
+        (shell.length should be > 0)(shell.length)
+      case ArgvCommand(argv) =>
+        (argv.size should be >0)(argv.size)
+    }
+  }
+
+  def healthCheckValidator(endpoints: Seq[Endpoint]) = validator[HealthCheck] { hc =>
+    hc.gracePeriodSeconds should be >= 0L
+    hc.intervalSeconds should be >= 0
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala]: What does interval 0 mean?

File: src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala [hunk 2]
```diff
@@ -64,9 +65,84 @@ trait PodsValidation {
...
+    hc.maxConsecutiveFailures should be >= 0
+    hc.timeoutSeconds should be >= 0
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala]: timeout 0 is probably wrong. Also: the timeout should be less than the interval.

File: src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala [hunk 3]
```diff
@@ -64,9 +65,84 @@ trait PodsValidation {
...
+    hc.delaySeconds should be >= 0
+    hc.http is optional(httpHealthCheckValidator(endpoints))
+    hc.tcp is optional(tcpHealthCheckValidator(endpoints))
+    hc.command is optional(commandCheckValidator)
+    hc is isTrue("Only one of http, tcp, or command may be specified") { hc =>
+      hc.http.isDefined ^ hc.tcp.isDefined ^ hc.command.isDefined
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala]: true ^ true ^ true ==> true
If all 3 are defined this is true.

File: src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala [hunk 4]
```diff
@@ -64,9 +65,84 @@ trait PodsValidation {
...
+    }
+  }
+
+  val endpointValidator = validator[Endpoint] { endpoint =>
+    endpoint.name.length is between(1, 63)
+    endpoint.name should matchRegexFully(NamePattern)
+    endpoint.containerPort.getOrElse(1) is between(1, 65535)
+    endpoint.hostPort.getOrElse(0) is between(0, 65535)
+    endpoint.protocol.size should be > 0
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala]: I think we need to restrict the protocol strings to `tcp` or `udp`

File: src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala [hunk 5]
```diff
@@ -64,9 +65,84 @@ trait PodsValidation {
...
+  }
+
+  val imageValidator = validator[Image] { image =>
+    image.id.length is between(1, 1024)
+  }
+
+  // TODO(PODS): don't we need to know what path we're mounting, not just where it is in the container?
+  // kind of confused by this one.
+  @SuppressWarnings(Array("UnusedParameter"))
+  def volumeMountValidator(volumes: Seq[Volume]): Validator[VolumeMount] = validator[VolumeMount] { volumeMount => // linter:ignore:UnusedParameter
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala]:
```scala
volumeMount.name is isTrue("Referenced Volume in VolumeMount does not exist") { volumeName =>
      volumes.exists(_.name == volumeName)
}
```

File: src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala [hunk 6]
```diff
@@ -64,9 +65,84 @@ trait PodsValidation {
...
+    artifact.destPath.map(_.length).getOrElse(1) is between(1, 1024)
+  }
+
+  val lifeCycleValidator = validator[Lifecycle] { lc =>
+    lc.killGracePeriodSeconds.getOrElse(0.0) should be > 0.0
+  }
+
+  def containerValidator(volumes: Seq[Volume]): Validator[PodContainer] = validator[PodContainer] { container =>
     container.resources is valid(resourceValidator)
+    container.endpoints is empty or every(endpointValidator)
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala]: `every` checks every object in the seq. Is there additional value in `is empty or` ?

File: src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala [hunk 7]
```diff
@@ -64,9 +65,84 @@ trait PodsValidation {
...
+    container.image.getOrElse(Image(ImageType.Docker, "abc")) is valid(imageValidator)
     container.environment is optional(envValidator)
+    container.healthCheck is optional(healthCheckValidator(container.endpoints))
+    container.volumeMounts is empty or every(volumeMountValidator(volumes))
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala]: remove `is empty or`

File: src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala [hunk 8]
```diff
@@ -64,9 +65,84 @@ trait PodsValidation {
...
+    container.artifacts is empty or every(artifactValidator)
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala]: remove `is empty or`

**Comment by aquamatthias** [comment]
I would like to see some tests for this new logic.

**Comment by jdef** [comment]
+1 for tests

also needed:
- endpoint names should be unique within the context of the pod
- volume names should be unique within the context of the pod
- container names should be unique within the context of the pod
- lots of duplication re: name length and pattern matching; can't we consolidate this?

File: src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala (hunk 1 above)
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala]: I actually only added the validation rules that were specified in the raml across the board :/

File: src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala (hunk 2 above)
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala]: Timeout >= 0? Agree on interval. I should have read all of the raml comments on not just the raml restrictions.

File: src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala (hunk 3 above)
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala]: Doh.

File: src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala (hunk 5 above)
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala]: That I missed entirely when reading the raml comments

File: src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala (hunk 6 above)
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala]: Oh wix, so confusing. What's "empty" for?

**Comment by jasongilanfarr** [comment]
Can add more validation, can't realistically remove a lot of the duplication unless we restrict to a new type too, eg, raml.Port, raml.PortAllowEmphemeral, raml.PathId, etc.

File: src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala (hunk 4 above)
Comment (jdef) [inline src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala]: pretty sure that mesos doesn't restrict this. why would we?

**Comment by jasongilanfarr** [comment]
Because those are the only two protocols that actually do anything?

**Comment by jdef** [comment]
seems like an artificial restriction. what if someone wanted to specify
http or sctp or ... whatever. why should marathon decide what's appropriate
here?

**Comment by jasongilanfarr** [comment]
Because those are L7 protocols built on L4 protocols and this is as far as
I'm aware a L4 protocol which is only tcp or udp and related to the cgroup
restrictions.

**Comment by jdef** [comment]
this is nothing to do with cgroups. protocol is used for discovery,
possibly port mapping, whatever. it's not used for isolation or allocation

**Comment by jasongilanfarr** [comment]
Port mapping is almost definitely L4 only, are you positive we shouldn't be
restricting it to valid L4 protocols?

**Comment by jdef** [comment]
Not even mesos restricts this:
https://github.com/apache/mesos/blob/94189ee48b8d36c6f395a5b9fb7d8b92873e0802/include/mesos/mesos.proto#L1986

**Comment by jasongilanfarr** [comment]
I see that, but does it work as intended with arbitrary strings? Let's get confirmation.

**Comment by jdef** [comment]
It's up to some mesos module to deal with port mapping / discovery info.
AFAIK arbitrary strings were used on purpose so that module impls wouldn't
be restricted.

**Pull request merged**